### PR TITLE
Be more opinionated for setting up a dev env.

### PR DESCRIPTION
**Pull request by anntzer**
Tell users to always use a venv and always install in editable mode.

If they know better, great.  But let's not offer too many confusing
options to newcomers.

xref https://discourse.matplotlib.org/t/running-tests-on-git-master/20761/

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: doc/devel/contributing.rst [hunk 1]
```diff
@@ -115,28 +100,15 @@ reflected the next time you import the library.  If you change the
...
 
-Alternatively, if you do ::
-
-   python -mpip install -v .
-
-all of the files will be copied to the installation directory however,
-you will have to rerun this command every time the source is changed.
-Additionally you will need to copy :file:`setup.cfg.template` to
-:file:`setup.cfg` and edit it to contain ::
-
```
Comment (jklymak) [inline doc/devel/contributing.rst]: So is this not needed anymore?

Comment (anntzer) [reply doc/devel/contributing.rst]: Not if you always use an editable install.

**Comment by efiring** [comment]
I think that ignoring the conda alternative in this context is a mistake, and will lead to more confusion, not less.

**Comment by anntzer** [comment]
You can use venvs within a conda env (in fact this is my normal everyday setup on my work machine...).

**Comment by story645** [comment]
Agree with @efiring here. I don't think I've used a venv since before conda was a thing & I'd be majorly frustrated/confused with making a venv in my conda env

**Comment by efiring** [comment]
For a conda user the disadvantage of venv is that it is yet another arcane and peripheral thing to have to learn and keep straight.  What is the advantage of using one within a conda env rather than just using another conda env?

**Comment by anntzer** [comment]
I'm not dead set against it, someone just needs to write the instructions on how to set them up and activate them.

File: doc/devel/contributing.rst (hunk 1 above)
Comment (efiring) [reply doc/devel/contributing.rst]: Are pip and setuptools now smart enough to handle changes to extension code in editable mode?  My sense is that editable mode works great for changes to existing python modules, but not beyond that.

Comment (anntzer) [reply doc/devel/contributing.rst]: C files are mentioned above and still have the note "rebuild if you edit one of them".

**Comment by timhoffm** [comment]
> someone just needs to write the instructions on how to set them up and activate them.

Will do.

**Comment by jklymak** [comment]
@timhoffm I think @efiring was also tasked with adding `conda`, so maybe co-ordinate w/ him?

I'm tempted to merge as an improvement, and then let new changes be on top?

**Comment by h-vetinari** [comment]
No comments on the PR per se (except to agree with @efiring's [comment](https://github.com/matplotlib/matplotlib/pull/15961#issuecomment-566704556) above), but I'm wondering why dev workflow items are milestoned for the already much-delayed 3.2.0? This is easily something that could be added later.

**Comment by story645** [comment]
@h-vetinari mpl practice is to usually milestone against the next release & if it isn't release critical than it gets remilestoned before the release is cut.

**Comment by nschloe** [comment]
The instructions use the `python` command which is not available to most Linux users by default from next year on. Replacing it with `python3` wouldn't hurt.

**Comment by nschloe** [comment]
When following the new instructions on current master, I'm getting
```
[...]
lib/matplotlib/docstring.py:40: in __call__
    func.__doc__ %= self.params
E   KeyError: 'scale'
```
when running `pytest`.

**Comment by anntzer** [comment]
switching to py3 seems fine, done.
the error re: docstrings still looks like a messed up dev environment.

**Comment by timhoffm** [comment]
Is it really true, that calling `python` does not work anymore on linux, once Python 2 is removed? I can't believe that. On my machine `python` is just a link to `python2`. I'd assume that that link will point to `python3` in the future.

**Comment by anntzer** [comment]
This is explicitly allowed by https://www.python.org/dev/peps/pep-0394/#for-python-runtime-distributors.

**Comment by nschloe** [comment]
> Is it really true, that calling python does not work anymore on linux, once Python 2 is removed?

Well, there is the Debian policy https://www.debian.org/doc/packaging-manuals/python-policy/ch-python.html#s-base:

> The python binary package must also ensure that /usr/bin/python is provided, as a symlink to the current python2.Y executable.

Perhaps, at some point in the future, `/usr/bin/python` will be something other than python2, but it won't be Jan 1, 2020. Since mpl only supports Python3 from then on, we better be specific.

**Comment by timhoffm** [comment]
@anntzer in that case, please go through the whole file and replace all occurrences of `python`.

**Comment by anntzer** [comment]
Once you're in the env you can use python (the env will provide that unsuffixed name for you); the suffixed name is only necessary to set up the env.

**Comment by nschloe** [comment]
Another issue: `pytest` is not installed in the venv, so it will take the one from system. That's probably not what you want.

**Comment by anntzer** [comment]
ah yes, that can be fixed with the usual dance of `python -mpytest`.

**Comment by nschloe** [comment]
Nice. More: `python setup.py build_ext --inplace` doesn't install the dependencies, so when running pytest, you get a lot of import errors.

**Comment by anntzer** [comment]
You're supposed to run pip install -ve . at least once, and setup.py build_ext -i only on later runs.

**Comment by nschloe** [comment]
I didn't catch that from the new docs. Perhaps wording should be something like
```
[...]
C-extension source (which might happen if you change branches) you will need to run

    python -mpip install -ve .

or run `python setup.py build_ext --inplace` if you've done that at least once before.
```

**Comment by anntzer** [comment]
It's in the paragraph immediately above the edit? (you may want to expand the unmodified part of the file in the diff view)

**Comment by nschloe** [comment]
You're right, I had missed that when looking at the diff. :+1:

**Comment by nschloe** [comment]
Next: When running pytest, I'm getting lots of file-not-found errors like
```
E           FileNotFoundError: [Errno 2] No such file or directory: '/path/to/matplotlib/result_images/test_arrow_patches/fancyarrow_test_image-expected_svg.png'
```
The directory contains
```
fancyarrow_test_image-expected_pdf.png
fancyarrow_test_image-expected.png
fancyarrow_test_image-expected.svg
fancyarrow_test_image.pdf
fancyarrow_test_image_pdf.png
fancyarrow_test_image.png
fancyarrow_test_image.svg
```

**Comment by anntzer** [comment]
Can you please provide a full log of what you did, starting from setting up a clean venv?

**Comment by nschloe** [comment]
> Can you please provide a full log of what you did, starting from setting up a clean venv?

I get something else now, no idea why it's different. Still an error, though:
```
$ ls ~/matplotlib-env/
ls: cannot access '~/matplotlib-env/': No such file or directory
$ python3 -mvenv ~/matplotlib-env/
$ source ~/matplotlib-env/bin/activate
$ python -mpip install -ve .
[...]
Successfully installed cycler-0.10.0 kiwisolver-1.1.0 matplotlib numpy-1.17.4 pillow-6.2.1 pyparsing-2.4.5 python-dateutil-2.8.1 six-1.13.0
Cleaning up...
Removed build tracker '/tmp/pip-req-tracker-aczpokn_'
$ pip install pytest
[...]
Successfully installed attrs-19.3.0 importlib-metadata-1.3.0 more-itertools-8.0.2 packaging-19.2 pluggy-0.13.1 py-1.8.0 pytest-5.3.2 wcwidth-0.1.7 zipp-0.6.0
$ python -mpytest --maxfail=1
[...]
E           FileNotFoundError: [Errno 2] No such file or directory: '/path/to/source/result_images/test_arrow_patches/fancyarrow_test_image-expected_svg.png'
```

**Comment by anntzer** [comment]
Strange, can't repro locally.

What's `matplotlib.__version__`?  I assume this is a clean fresh clone?  Do you have inkscape installed?  What's the traceback on the FileNotFoundError (`pytest --full-trace` may help)?

**Comment by nschloe** [comment]
> What's matplotlib.__version__? 
```
$ python -c "import matplotlib; print(matplotlib.__version__)"
3.1.1.post2790+gce7115d3d
```

> I assume this is a clean fresh clone?
Indeed.

> Do you have inkscape installed?
```python
$ inkscape --version
Inkscape 1.0beta1 (fe3e306978, 2019-09-17)
```

The weird thing is that the directory contains all kinds of files,
```
fancyarrow_test_image-expected_pdf.png
fancyarrow_test_image-expected.png
fancyarrow_test_image-expected.svg
fancyarrow_test_image.pdf
fancyarrow_test_image_pdf.png
fancyarrow_test_image.png
fancyarrow_test_image.svg
```
but _not_ `fancyarrow_test_image-expected_svg.png`. It appears some renaming issue.

Well, the issue seems unrelated to this PR and I don't want to delay it.

**Comment by anntzer** [comment]
Ah yes, I *think* inkscape 1.0 changed their command line interface and we'll probably need to fix some stuff on mpl's side to support that, can you downgrade to inkscape<1.0 for now?

**Comment by nschloe** [comment]
> can you downgrade to inkscape<1.0 for now?

I wouldn't know how. (My inkscape comes from the Ubuntu archives.)

**Comment by nschloe** [comment]
Okay, got the test to pass now (after https://github.com/matplotlib/matplotlib/pull/16002). LGTM.

**Comment by timhoffm** [comment]
Even without conda, this is an improvement. Conda-based configuration can follow in a separate PR.

**Comment by timhoffm** [comment]
Backport would need at least also #15476 and #15569. Maybe more.

**Comment by timhoffm** [comment]
Giving up on backporting, because the local freetype change #15476 won‘t be in 3.2.

**Pull request merged**